### DISCUSSION TITLE
Want to understand the issue with BPFTRACE_HAVE_BTF

### DISCUSSION THREAD

**Discussion by Engg-blogger** [bpftrace/bpftrace#3054] [General]
Hi Team - I have deployed latest version of bpftrace (appimage) in ubuntu 20.04. I am running `tcpconnect.bt` from tools section. 
But while running the` tcpconnect.bt`, it gives me an error that `sys/socket.h` is not available. 
I was looking at the code for `tcpconnect.bt` and it has following include files:

```
#ifndef BPFTRACE_HAVE_BTF
#include <linux/socket.h>
#include <net/sock.h>
#else
#include <sys/socket.h>
#endif
```
I also ran the script check_kernel_features.sh which returned that all features are available. Linux headers are also available. 
Could you explain why the `sys/socket.h` is being called when BTF is enabled on the VM? What could be missing?

**Comment by viktormalik** [comment]
tcpconnect.bt uses the `AF_INET` and `AF_INET6` macros which are defined in `linux/socket.h` but also in a userspace header `sys/socket.h` (should be in glibc headers). Since BTF doesn't contain macro definitions, we chose to include the userspace header as its definitions don't conflict with BTF definitions and therefore can be used simultaneously (which wouldn't be the case if we always included `linux/socket.h`).

TLDR; you need to install the glibc-headers package (or other package containing `sys/socket.h`).

**Comment by jordalgo** [comment]
I added a little comment to this file, which may help future users: https://github.com/bpftrace/bpftrace/pull/3055

**Comment by ajor** [comment]
If the question is "why is sys/socket.h used instead of linux/socket.h":
The first line of the if-macro is is an "if-n-def" (i.e. if-not-defined), not "if-def", so when BTF is detected on your system, sys/socket.h will be used.

**Reply by Engg-blogger** [reply]
Ok.. that answers my question. Do u have a minimum version compatibility for glibc required to run tcp scripts with bpftrace v0.20.0?

**Reply by ajor** [reply]
glibc's very stable, so I don't think it will have changed enough any time recently for older versions to be incompatible with bpftrace

**Reply by Engg-blogger** [reply]
Ok. So, I downloaded glibc-source and libc6 on my ubuntu VM, but still getting the same error that sys/socket.h file not found. So, I downloaded gcc and it started working. So, does that mean I need to have gcc on my machine for tcp_connect to work?

**Reply by viktormalik** [reply]
What distro is this?

**Reply by Engg-blogger** [reply]
Ubuntu 20.04

**Reply by Engg-blogger** [reply]
Also, another question is if I can add gcc as part of appimage. Will that suffice?

**Reply by viktormalik** [reply]
On Ubuntu, the file is a part of the `libc6-dev` package. The package is a dependency of gcc but you don't need to install gcc, just `libc6-dev`.
